### PR TITLE
Added the SearchPostsInTeam method to the plugin API

### PR DESCRIPTION
**Pull request by awbraunstein**
#### Summary
Adds the `SearchPostsInTeam` method to the plugin API.

This spits the old `App.SearchPostsInTeam` function into two functions: `App.SearchPostsInTeamForUser` which is the old function and `App.SearchPostsInTeam` which is a new function that no longer requires the current `userId` to be provided.

This means that for searches that would involve channels between users, all users must be specified.
#### Ticket Link
https://github.com/mattermost/mattermost-server/issues/10099

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Added API documentation (required for all new APIs)
- [x] All new/modified APIs include changes to the drivers

### CODE REVIEW

File: store/sqlstore/post_store.go [hunk 1]
```diff
@@ -267,14 +267,14 @@ func (s *SqlPostStore) GetFlaggedPostsForChannel(userId, channelId string, offse
 
 		var posts []*model.Post
 		query := `
-			SELECT 
-				* 
-			FROM Posts 
-			WHERE 
-				Id IN (SELECT Name FROM Preferences WHERE UserId = :UserId AND Category = :Category) 
+			SELECT
```
[the remark this answers is not in the document]
Comment (awbraunstein) [reply store/sqlstore/post_store.go]: Can do! Just for future reference, is there a reason for this whitespace?

File: store/sqlstore/post_store.go [hunk 2]
```diff
@@ -1257,11 +1262,11 @@ func (s *SqlPostStore) determineMaxPostSize() int {
 		// The Post.Message column in MySQL has historically been TEXT, with a maximum
 		// limit of 65535.
 		if err := s.GetReplica().SelectOne(&maxPostSizeBytes, `
-			SELECT 
+			SELECT
```
[the remark this answers is not in the document]
Comment (awbraunstein) [reply store/sqlstore/post_store.go]: Removed.

File: plugin/api.go [hunk 3]
```diff
@@ -221,6 +221,11 @@ type API interface {
 	// Minimum server version: 5.6
 	SearchUsers(search *model.UserSearch) ([]*model.User, *model.AppError)
 
+	// SearchPostsInTeam returns a list of posts in a specific team that match the given params.
+	//
+	// Minimum server version: TODO: Fill in minimum version.
```
[the remark this answers is not in the document]
Comment (awbraunstein) [reply plugin/api.go]: Great! I'll update the comment.

File: app/plugin_api.go [hunk 4]
```diff
@@ -338,6 +338,37 @@ func (api *PluginAPI) SearchUsers(search *model.UserSearch) ([]*model.User, *mod
 	return api.app.SearchUsers(search, pluginSearchUsersOptions)
 }
 
+func (api *PluginAPI) SearchPostsInTeam(teamId string, params *model.SearchParameter) (*model.PostList, *model.AppError) {
```
[the remark this answers is not in the document]
Comment (awbraunstein) [reply app/plugin_api.go]: There are two competing proposals in my mind. The first being what the original bug was suggesting which was basically building something that is analogous to the public API for searching posts and the second being essentially a frontend for the database query. It seems like what we actually want is the latter. Is that correct?

File: store/sqlstore/post_store.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (crspeller) [reply store/sqlstore/post_store.go]: No I don't think so. If you want to submit a separate PR to fix it up, that would be nice.

File: app/plugin_api.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (crspeller) [reply app/plugin_api.go]: Yes. One of the guiding principals for plugins is to give them as much power as possible.

[the remark this answers is not in the document]
Comment (awbraunstein) [reply app/plugin_api.go]: I updated this to be more of a frontend for the db.

**Comment by awbraunstein** [comment]
@crspeller I have updated this PR with the changes from your comments. PTAL

File: plugin/api.go [hunk 5]
```diff
@@ -221,6 +221,11 @@ type API interface {
 	// Minimum server version: 5.6
 	SearchUsers(search *model.UserSearch) ([]*model.User, *model.AppError)
 
+	// SearchPostsInTeam returns a list of posts in a specific team that match the given params.
+	//
+	// Minimum server version: 5.10
+	SearchPostsInTeam(teamId string, paramsList []*model.SearchParams) (*model.PostList, *model.AppError)
```
Comment (hanzei) [inline plugin/api.go]: Could you please change the return value to `([]*model.Post, *model.AppError)`? `model.PostList` is more of an internal type.

File: app/post.go [hunk 6]
```diff
@@ -744,7 +744,44 @@ func (a *App) parseAndFetchChannelIdByNameFromInFilter(channelName, userId, team
 	return channel, nil
 }
 
-func (a *App) SearchPostsInTeam(terms string, userId string, teamId string, isOrSearch bool, includeDeletedChannels bool, timeZoneOffset int, page, perPage int) (*model.PostSearchResults, *model.AppError) {
+type paramsModifier func(*model.SearchParams)
```
Comment (hanzei) [inline app/post.go]: Do you really need to type this?

File: model/search_params.go [hunk 7]
```diff
@@ -23,6 +23,8 @@ type SearchParams struct {
 	OrTerms                bool
 	IncludeDeletedChannels bool
 	TimeZoneOffset         int
+	// True iff this search doesn't originate from a "current user".
```
Comment (hanzei) [inline model/search_params.go]:
```suggestion
	// True if this search doesn't originate from a "current user".
```

Comment (awbraunstein) [reply model/search_params.go]: Done.

File: app/post.go (hunk 6 above)
Comment (awbraunstein) [reply app/post.go]: Not really. I removed it.

File: plugin/api.go (hunk 5 above)
Comment (awbraunstein) [reply plugin/api.go]: Done.

**Comment by crspeller** [comment]
@awbraunstein Looks like you need to regenerate the RPC code and rebase.

**Comment by awbraunstein** [comment]
@crspeller regenerated and rebased against master

**Comment by awbraunstein** [comment]
I'm all good on this PR if someone wants to merge it for me. Thanks!

**Comment by hanzei** [comment]
Thanks @awbraunstein for the contribution! :tada:

**Pull request merged**